cloud-init/write_disable_inactive_users_cron_job.tpl.yml

```diff
@@ -0,0 +1,88 @@
+---
+# Write a shell script to disable inactive FreeIPA users.  This cron
+# job will be run daily.
+write_files:
+  - path: /etc/cron.daily/disable_inactive_users.sh
+    permissions: "0500"
+    owner: root:root
+    content: |
+      #!/usr/bin/env bash
+
+      # Input variables are:
+      # domain - the domain for the IPA server (e.g. example.com)
+
+      set -o nounset
+      set -o errexit
+      set -o pipefail
+
+      ###
+      # We're running as root, so login to kerberos using the host's keytab
+      #
+      # It would be better to create a kerberos service instead, but
+      # that is left as a future improvement.
+      ###
+      kinit -k -t /etc/krb5.keytab
+
+      ###
+      # Convert the domain into an LDAP searchbase
+      ###
+      searchbase="cn=users,cn=accounts"
+      # The "domain" item below that looks like shell variable is actually
+      # replaced by the Terraform templating engine.  Hence we can ignore
+      # the "undefined variable" warnings from shellcheck.
+      #
+      # shellcheck disable=SC2154
+      g="${domain}"
+      # While g still contains a dot character...
+      while grep -F "." <<< "$g"
+      do
+          # Extract the longest non-dot-containing string from the beginning
+          # of $g
+          tmp=$(expr "$g" : '\([^\.]*\)')
+          # Append $tmp onto the end of $searchbase
+          searchbase=$searchbase,dc=$tmp
+          # Remove tmp and its trailing period from $g
+          #
+          # I'd like to do this more simply, but constructs with
+          # complicated braces seem to foul up the templating foo in
+          # ../cloud_init.tf.
+          length_tmp=$(expr length $tmp)
+          length_g=$(expr length $g)
+          g=$(expr substr $g $(($length_tmp + 2)) $length_g)
+      done
+      # There are no more dots in $g, so we just have to append the last bit
+      # of $g onto $searchbase
+      searchbase=$searchbase,dc=$g
+
+      ###
+      # Determine the date 45 days ago, in a format that ldapsearch likes
+      ###
+      distant_past=$(date --date="$(date) -45 days" +%Y%m%d%H%M%SZ)
+
+      ###
+      # Query LDAP to determine all users that have not logged in for 45 days
+      ###
+      users_to_disable=$(ldapsearch \
+                               -b "$searchbase" \
+                               "(krbLastSuccessfulAuth<=$distant_past)" \
+                               uid \
+                               2>/dev/null \
+                             | grep "^uid:" \
+                             | sed "s/^uid: \(.*\)/\1/")
+
+      ###
+      # Disable the users
+      ###
+      for user in $users_to_disable
+      do
+          # The ipa user-disable command may fail if the user is already
+          # disabled, but that's not really en error condition for us.
```

Comment:
Can you think of any other ways the `ipa user-disable` might fail that would be unintentionally masked by `set +o errexit`?  Is this something worth worrying about?

Reply:
I don't think it's something worth worrying about.  I can't think of other reasons right now (although I'm sure they exist) but if the `ipa user-disable` fails the reason why will be apparent in the cron logs.

All `set -o errexit` does is exit the bash script if any command returns an exit code indicating a failure.  We're temporarily turning that behavior off so that the loop can continue on to disabling the next inactive user.

I'll add a comment clarifying this.

Reply:
See 174ffa3.

Reply:
Thanks- makes sense!